bin/i18n/curriculum_sync_utils.rb

```diff
@@ -0,0 +1,204 @@
+require 'cdo/languages'
+
+require_relative 'curriculum_sync_utils/serializers'
+
+# This module contains all i18n sync logic specific to the content
+# imported into Code Studio from CurriculumBuilder; Lessons, Activities,
+# Resources, Vocabularies, Objectives, etc.
+module CurriculumSyncUtils
+  def self.sync_in
+    puts "Sync in curriculum content"
+    sync_in_serialize
+    # TODO: redaction
+  end
+
+  def self.sync_out
+    puts "Sync out curriculum content"
+    # TODO: restoration
+    sync_out_reorganize
+  end
+
+  # Helper method to get the desired destination subdirectory of the given
+  # script for the sync in. Note this may be a nested directory like "2021/csf"
+  def self.get_script_subdirectory(script)
+    # special-case Hour of Code scripts.
+    return "Hour of Code" if ScriptConstants.script_in_category?(:hoc, script.name)
+
+    # catchall for scripts without courses
+    return 'other' unless script.get_course_version.present?
+
+    # special-case CSF; we want to group all CSF courses together, even though
+    # they all have different course offerings.
+    return File.join(script.get_course_version.key, 'csf') if script.csf?
+
+    # base case
+    return File.join(script.get_course_version.key, script.get_course_version.course_offering.key)
+  end
+
+  # Serialize all curriculum data to the I18N source directory.
+  #
+  # Relies heavily on the custom serializer logic defined in
+  # curriculum_sync_utils/serializers to generate a hash of results with
+  # translator-readable keys, rather than the basic array that is produced by
+  # default.
+  def self.sync_in_serialize
+    Script.all.each do |script|
+      # TODO: what else do we want to consider when deciding what to sync?
+      next unless script.is_migrated?
+
+      # prepare data
+      data = ScriptCrowdinSerializer.new(script).as_json.compact
+      data.delete(:crowdin_key) # don't need this for top-level data
+
+      next unless data.present?
+
+      # write data to path
+      # TODO: include the "other directory already exists" logic from localize_level_content
+      path = File.join(I18N_SOURCE_DIR, 'curriculum_content', get_script_subdirectory(script), "#{script.name}.json")
+      FileUtils.mkdir_p(File.dirname(path))
+      File.write(path, JSON.pretty_generate(data))
+    end
+  end
+
+  # Given a deeply-nested hash of serialized objects (as generated by a
+  # CrowdinCollectionSerializer) and a Serializer for the top-level object,
+  # recursively extract all nested objects and return a flattened hash
+  # organized by type.
```

Comment:
.script_json is flat. the input hash is deeply-nested. the output hash is flat. can you comment on why the input hash should be deeply nested? "convenience" is fine, but I don't think I understand the code well enough yet to see the convenience.

Reply:
We aren't using script_jsons here. The input hash is deeply-nested because the actual conceptual structure of the content itself is deeply nested